### PR TITLE
隠れ状態の線形補間

### PR DESCRIPTION
**Pull request by myxyy**
## 概要

隠れ状態チャネル毎にGlobalConvolutionしていたのを
隠れ状態を線形変換してから重みづけ足し合わせるよう変更

SioConv4段は流石に少ないので8段にしてます

## Submit前の確認項目

- [x] タイトルは一目でわかるようにし、説明文は PR を簡潔に説明するようにしましたか?
- [x] あなたの PR が、異なる変更を束ねたものではなく、ひとつのことだけを行うものであることを確認しましたか?
- [x] この PR で導入されたすべての変更点をリストアップしましたか?
- [x] PR を`make run`コマンドでローカルでテストしましたか?

### CODE REVIEW

File: ami/models/policy_value_common_net.py [hunk 1]
```diff
@@ -70,3 +70,60 @@ def forward(self, flattened_obs: Tensor, stacked_hidden: Tensor) -> Tensor:
...
+
+        Return shape: (*, N)
+    """
+
+    def __init__(self, dim: int, depth: int, num_head: int) -> None:
+        super().__init__()
+        self.hidden_proj = nn.Parameter(torch.randn(depth, dim, dim) * (dim**-0.5))
+        self.logit_coef_proj = nn.Linear(depth * dim, depth)
+        self.num_head = num_head
+        self.norm = nn.GroupNorm(num_head, num_head)
```
Comment (Geson-anko) [inline ami/models/policy_value_common_net.py:89]: GroupNormじゃなくて InstanceNormじゃだめなんですか？
https://pytorch.org/docs/stable/generated/torch.nn.GroupNorm.html

Comment (myxyy) [reply ami/models/policy_value_common_net.py]: https://github.com/MLShukai/ami-vconf24/pull/22/commits/ac868afdb1dc21aa7ea0dd70ff67fd6ef5e11cc3
InstanceNormにしました

File: ami/models/policy_value_common_net.py [hunk 2]
```diff
@@ -70,3 +70,60 @@ def forward(self, flattened_obs: Tensor, stacked_hidden: Tensor) -> Tensor:
...
+        )
+
+        logit_coef = self.logit_coef_proj(stacked_hidden.reshape(batch, depth * dim))
+
+        out = torch.einsum(
+            "bd,dij,bdj->bi",
+            nn.functional.softmax(logit_coef, dim=-1).squeeze(-1),
+            self.hidden_proj,
+            stacked_hidden,
+        )
```
Comment (Geson-anko) [inline ami/models/policy_value_common_net.py:108]: depth方向それぞれのhiddenに対して、nn.Linear(N,N)の操作をすると思っていたのですが、これだとバイアス項が全てのhiddenに対して0で統一される実装になりますが、それは大丈夫ですか？
`einsum`記法はあまり浸透仕切っていないので、コメントなどをつけるか、もっとわかりやすい実装にしていただけると助かります。

File: configs/experiment/world_models_sioconv_lerp_hidden.yaml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  ppo_trajectory:
+    use_embed_obs_as_observation: True
+  forward_dynamics_trajectory:
+    max_len: ${python.eval:"(128 + 1) * 16"} # batch_size * 16
+
+trainers:
+  forward_dynamics:
+    partial_dataloader:
+      batch_size: ${python.eval:128+1}
+    minimum_new_data_count: ${.partial_dataloader.batch_size}
```
Comment (Geson-anko) [inline configs/experiment/world_models_sioconv_lerp_hidden.yaml:29]: `world_models_sioconv`のexperiment configをそのまま引き継ぐこともできます。
設定ファイルの依存関係が増えて追いにくくなるデメリットはありますが、一応楽に書けるので例として提示します。
```suggestion
defaults:
  - world_models_sioconv
  - override /models: world_models_sioconv_lerp_hidden
```

File: configs/models/world_models_sioconv_lerp_hidden.yaml [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+      _target_: torch.nn.Identity
+    policy_head:
+      _target_: ami.models.components.discrete_policy_head.DiscretePolicyHead
+      dim_in: ${..observation_hidden_projection.dim_out}
+      action_choices_per_category:
+        _target_: hydra.utils.get_object
+        path: ami.interactions.environments.actuators.vrchat_osc_discrete_actuator.ACTION_CHOICES_PER_CATEGORY
+    value_head:
+      _target_: ami.models.components.fully_connected_value_head.FullyConnectedValueHead
+      dim_in: ${..observation_hidden_projection.dim_out}
```
Comment (Geson-anko) [inline configs/models/world_models_sioconv_lerp_hidden.yaml:91]: よくよくみたらこのモデル活性化関数なくないですか。David HaのWorld Modelsだと確かに活性化関数なかったので良いですが、 `core_model`に`nn.ReLU()`でも指定しておいた方が良くないですか？
ちなみに Global Convの時は内部にReLUが入っていたのでnn.Identityを指定していました。

**Review by Geson-anko — COMMENTED** [review]
@myxyy 実装ありがとうございます。
Approveしても良いのですが、いくつか疑問点があるのでよろしくお願いします。

File: ami/models/policy_value_common_net.py [hunk 5]
```diff
@@ -70,3 +70,60 @@ def forward(self, flattened_obs: Tensor, stacked_hidden: Tensor) -> Tensor:
...
         return out
+
+
+class LerpStackedHidden(nn.Module):
+    """Linear interpolation along depth of stacked hidden.
+
+    Shape:
+        - stacked_hidden: (*, D, N)
+
+        Return shape: (*, N)
```
[the remark this answers is not in the document]
Comment (myxyy) [reply ami/models/policy_value_common_net.py]: https://github.com/MLShukai/ami-vconf24/pull/22/commits/36eec698f3316b5ee83ed84dea2871dc57fdfc34
修正しました
`ConcatFlattenedObservationAndStackedHidden`のコメントも修正必要でしょうか？

File: ami/models/policy_value_common_net.py (hunk 2 above)
Comment (myxyy) [reply ami/models/policy_value_common_net.py]: https://github.com/MLShukai/ami-vconf24/pull/22/commits/ee67532b0ce88c4f3644579ab02aa264946d3728
バイアス無くてもよいかと思いましたがバイアス付けて整理しました

File: configs/models/world_models_sioconv_lerp_hidden.yaml (hunk 4 above)
Comment (myxyy) [reply configs/models/world_models_sioconv_lerp_hidden.yaml]: https://github.com/MLShukai/ami-vconf24/pull/22/commits/b84ce344eb4ab36eaf1a70ee6080a0a975b145ba
せっかくなのでResNetを入れました

File: configs/experiment/world_models_sioconv_lerp_hidden.yaml (hunk 3 above)
Comment (myxyy) [reply configs/experiment/world_models_sioconv_lerp_hidden.yaml]: 省略しました https://github.com/MLShukai/ami-vconf24/pull/22/commits/7116aec26053c6470ae1bfa20aa9b05535fa03b1

File: ami/models/policy_value_common_net.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Geson-anko) [reply ami/models/policy_value_common_net.py]: @myxyy `ConcatFlattenedObservationAndStackedHidden`は問題ないので大丈夫です。

File: tests/models/components/test_resnet.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import pytest
+import torch
+
+from ami.models.components.resnet import ResNetFF
+
+
+class TestMultiEmbedding:
```
Comment (Geson-anko) [inline tests/models/components/test_resnet.py:7]: 名前ミス
```suggestion
class TestResNetFF:
```

**Review by Geson-anko — APPROVED** [review]
@myxyy 実装ありがとうございました。若干の名前ミスがあるのでそれ修正したらマージしてください。

File: ami/models/components/resnet.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import torch.nn as nn
+from torch import Tensor
+
+
+class ResNetFF(nn.Module):
+    def __init__(self, dim: int, dim_hidden: int, depth: int, activation: nn.Module = nn.ReLU()):
+        super().__init__()
+        self.ff_list = nn.ModuleList(
+            [nn.Sequential(nn.Linear(dim, dim_hidden), activation, nn.Linear(dim_hidden, dim)) for _ in range(depth)]
```
Comment (Geson-anko) [inline ami/models/components/resnet.py:9]: @myxyy 本家ResNetでは BatchNormとか入ってましたけど、Normalization Layer入れなくても大丈夫ですかね？

File: tests/models/components/test_resnet.py (hunk 6 above)
Comment (myxyy) [reply tests/models/components/test_resnet.py]: https://github.com/MLShukai/ami-vconf24/pull/22/commits/729d246fcfff73fc60c24b97a9a5a2515150d8ae
修正しました

File: ami/models/components/resnet.py (hunk 7 above)
Comment (myxyy) [reply ami/models/components/resnet.py:9]: とりあえず動いてるのでこのままで行きます
